### PR TITLE
Colin Dismuke - React-Todo

### PR DESCRIPTION
**Pull request by cpdis**
☑️ @John_Spraul
☑️ @John_Spraul
☑️ @John_Spraul

### CODE REVIEW

**Comment by cpdis** [comment]
[React Todo App on Netlify 🌎 ](https://colin-react-todo.netlify.com/)

File: src/App.js [hunk 1]
```diff
@@ -1,16 +1,158 @@
...
+addToDo = (newTask, id) => {
+  // Create a new todo object based on a given task and id
+  const todo = {
+    task: `${newTask}`,
+    id: id,
+    completed: false,
+  }
+
+  // Get the todo array and add the new todo to the end using push()
+  const todos = this.state.todos;
```
Comment (John-Spraul) [inline src/App.js]: this line should be `const todos = this.state.todos.slice()`, so you make a copy of the current todos in state, but don't actually modify it directly

File: src/App.js [hunk 2]
```diff
@@ -1,16 +1,158 @@
...
+}
+
+// Need a handler that takes the input from the form and creates new todo
+addToDoHandler = (element) => {
+
+  // This prevents the page from being refreshed, otherwise you lose everything when you submit
+  element.preventDefault();
+
+  // Get the input from the node immediately preceding the current one. See: https://developer.mozilla.org/en-US/docs/Web/API/Node/previousSibling
+  const input = element.target.previousSibling.value;
```
Comment (John-Spraul) [inline src/App.js]: interesting way of solving this problem

Go-to React approach is to hold the input value in state combined with a handling function that changes the value as the form/input is filled out

File: src/App.js [hunk 3]
```diff
@@ -1,16 +1,158 @@
...
+  
+  // Set the current date as the id for the todo
+  const id = Date.now();
```
Comment (John-Spraul) [inline src/App.js]: easy enough to assign this in the `addToDo` function, then you don't have to worry about passing in an second argument

File: src/App.js [hunk 4]
```diff
@@ -1,16 +1,158 @@
...
+
+  // Return the newly created object and set the state with it using addToDo function
+  return this.addToDo(input, id)
```
Comment (John-Spraul) [inline src/App.js]: why return if `addToDo` doesn't return anything?

File: src/App.js [hunk 5]
```diff
@@ -1,16 +1,158 @@
...
+    
+    // Get the todo array so that the completed state of the current todo can be toggled between true and false.
+    // Iterate through using .map
+    const todos = this.state.todos.map(todo => {
+
+      // Check if current todo matches the todo id
+      // Convert current.id to a number since it's currently a string
+      if (Number(current.id) === todo.id) {
+        // Add a class that strikes through the word
+        current.classList.toggle('completed');
```
Comment (John-Spraul) [inline src/App.js]: best to leave the class toggling stuff to React instead of using `classList.toggle`

File: src/App.js [hunk 6]
```diff
@@ -1,16 +1,158 @@
...
+
+        // Sets completed to whatever the opposite value is
+        todo.completed = !todo.completed;
+      }
+
+      // Just return the todo if it doesn't match the current target
```
Comment (John-Spraul) [inline src/App.js]: this comment is misleading, because the function returns either way

File: src/App.js [hunk 7]
```diff
@@ -1,16 +1,158 @@
...
+     // Set the current state with the new todo array
+    this.setState({todos})
+    localStorage.setItem("todos", JSON.stringify(todos));
+}
+
+// Need a function that removes completed todos
+removeComplete = (element) => {
+
+    // This prevents the page from being refreshed, otherwise you lose everything when you submit
+    element.preventDefault();
```
Comment (John-Spraul) [inline src/App.js]: `.preventDefault` is a method for events, so element might be the wrong placeholder choice

File: src/App.js [hunk 8]
```diff
@@ -1,16 +1,158 @@
...
+
+    // Create an empty array to hold incomplete items
+    let filtered = [];
+
+    // Get all of the todos with the completed attribute set to false
+    // Use something similar to toggleComplete but use const todos = this.state.todos.filter instead
+    let todos = this.state.todos.filter(todo => {
```
Comment (John-Spraul) [inline src/App.js]: this isn't the right way to use the `filter` method; you're using it as a `for` loop

File: src/App.js [hunk 9]
```diff
@@ -1,16 +1,158 @@
...
+      return filtered;
+    });
+
+    // Set the current state with the new todo array
+    todos = filtered;
+    this.setState({todos})
+    localStorage.setItem("todos", JSON.stringify(todos));
+}
+
+hydrateStateWithLocalStorage() {
```
Comment (John-Spraul) [inline src/App.js]: 💧

File: src/App.js [hunk 10]
```diff
@@ -1,16 +1,158 @@
...
+  // Iterate through all items in state
+  // This was extremely helpful: https://hackernoon.com/how-to-take-advantage-of-local-storage-in-your-react-projects-a895f2b2d3f2
+  for (let key in this.state) {
```
Comment (John-Spraul) [inline src/App.js]: what else are you storing in localStorage other than the todos array?

File: src/App.js [hunk 11]
```diff
@@ -1,16 +1,158 @@
...
+    // Check to see if the key exists in localStorage. See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty
+    if (localStorage.hasOwnProperty(key)) {
+      let value = localStorage.getItem(key);
+
+      // Parse the localStorage string and setState
+      try {
+        value = JSON.parse(value);
+        this.setState({ [key]: value })
+      } catch (e) {
```
Comment (John-Spraul) [inline src/App.js]: what is this `catch` supposed to do?

File: src/components/TodoComponents/TodoList.js [hunk 12]
```diff
@@ -1,2 +1,20 @@
...
+import Todo from './Todo';
+
+const TodoList = (props) => {
+    return (
+      <div className="list-container">
+          {/* Takes as input (from props) each of the todos in state, iterates through (using .map), and renders them onscreen */}
+          {props.todos.map(todo => {
+            return (
+              <div>
+                <Todo key={todo.id} id={todo.id} task={todo.task} onClick={props.toggleComplete} completed={todo.completed} />
```
Comment (John-Spraul) [inline src/components/TodoComponents/TodoList.js]: you're passing down parts of `todo` as props individually, why not just pass down the whole `todo` object?

**Pull request closed without merging**